src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs

```diff
@@ -244,6 +244,85 @@ public void TestFailedLoadDoesntCauseNoAnalyzersWarning()
         }
 
+        [Fact]
+        [WorkItem(1032909, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/1032909")]
+        public void TestReferencingFakeCompiler()
+        {
+            AnalyzerFileReference reference = CreateAnalyzerFileReference(_testFixture.AnalyzerWithFakeCompilerDependency.Path);
+
+            List<AnalyzerLoadFailureEventArgs> errors = new List<AnalyzerLoadFailureEventArgs>();
+            EventHandler<AnalyzerLoadFailureEventArgs> errorHandler = (o, e) => errors.Add(e);
+            reference.AnalyzerLoadFailed += errorHandler;
+            var builder = ImmutableArray.CreateBuilder<DiagnosticAnalyzer>();
+            reference.AddAnalyzers(builder, LanguageNames.CSharp);
+            var analyzers = builder.ToImmutable();
+            reference.AnalyzerLoadFailed -= errorHandler;
+
+            Assert.Equal(1, errors.Count);
+            var error = errors[0];
+
+            // failure is in the analyzer itself, i.e. abstract members on DiagnosticAnalyzer are not implemented.
+            Assert.Equal(AnalyzerLoadFailureEventArgs.FailureErrorCode.UnableToCreateAnalyzer, error.ErrorCode);
+            Assert.Equal("Analyzer", error.TypeName);
+        }
+
+        [Fact]
+        [WorkItem(1032909, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/1032909")]
+        public void TestReferencingLaterFakeCompiler()
+        {
+            AnalyzerFileReference reference = CreateAnalyzerFileReference(_testFixture.AnalyzerWithLaterFakeCompilerDependency.Path);
+
+            List<AnalyzerLoadFailureEventArgs> errors = new List<AnalyzerLoadFailureEventArgs>();
+            EventHandler<AnalyzerLoadFailureEventArgs> errorHandler = (o, e) => errors.Add(e);
+            reference.AnalyzerLoadFailed += errorHandler;
+            var builder = ImmutableArray.CreateBuilder<DiagnosticAnalyzer>();
+            reference.AddAnalyzers(builder, LanguageNames.CSharp);
+            var analyzers = builder.ToImmutable();
+            reference.AnalyzerLoadFailed -= errorHandler;
+
+            Assert.Equal(1, errors.Count);
+            var error = errors[0];
+            Assert.Equal(AnalyzerLoadFailureEventArgs.FailureErrorCode.ReferencesNewerCompiler, error.ErrorCode);
+            Assert.Null(error.TypeName);
+        }
+
+        private class AnalyzerLoaderMockCSharpCompiler : CSharpCompiler
+        {
+            public AnalyzerLoaderMockCSharpCompiler(CSharpCommandLineParser parser, string? responseFile, string[] args, BuildPaths buildPaths, string? additionalReferenceDirectories, IAnalyzerAssemblyLoader assemblyLoader, GeneratorDriverCache? driverCache = null, ICommonCompilerFileSystem? fileSystem = null)
+                : base(parser, responseFile, args, buildPaths, additionalReferenceDirectories, assemblyLoader, driverCache, fileSystem)
+            {
+            }
+        }
+
+        [ConditionalFact(typeof(IsEnglishLocal))]
+        public void AssemblyLoading_ReferencesLaterFakeCompiler_EndToEnd_CSharp()
```

Comment:
I also wanted to do an end-to-end test for VB, but I wasn't able to figure out how to provide a ref to Microsoft.VisualBasic.dll and System.dll in this context. Tried a few ways of making a mock compiler.